### ISSUE TITLE
[Rendering] RendererRD allocates excessive RIDs

### ISSUE THREAD

**Issue by AThousandShips** [godotengine/godot#89680]
### Tested versions

4.x, at least in 4.2 and later, haven't checked code in 4.1 and prior

### System information

N/A

### Issue description

As seen in:
* https://github.com/godotengine/godot/issues/89602

The renderer (Forward+ and Mobile only) allocates a lot of RIDs for seemingly simple operations, this isn't necessarily a problem but at least to me the general assumption is that RIDs won't be allocated unless you create or update some resource, adding or changing surfaces for meshes, updating shaders, etc. makes sense but in this case it feels excessive

While the specific case above might be considered extreme it isn't obvious that there's a maximum number of polygons (or circles, or other things drawn with a polygon as backing) you can draw ever, until you hit RID exhaustion, at first I couldn't even figure out what was wrong in that case as it shouldn't have been exhausting these

I'm not sure how to best solve this, but seeing how the data here is strictly internal to the renderer, and the RIDs won't pass to the user, I'd say it would be safe to re-use the RIDs and have a pool of them for drawing (i.e. reuse the storage, instead of freeing and allocating), if some alternative method for doing this entirely can't be done and avoid allocations entirely, I'm not fully familiar with the low level details on the ways to allow reusing these etc., but I imagine it should be possible 

I was trying to figure out how to solve this but the complexity of updating the data correctly was a bit too complex for my current understanding of the internals so I though I'd bring it to broader attention to discuss how to improve on it

Tl;Dr; the renderer allocates 4 RIDs every time you draw a circle, this exhausts them and makes just drawing a circle an action you can do a finite number of times, this is unexpected and hard to manage

To put this into perspective:
If you draw just 100 circles every frame, and render at 120 FPS, it takes just 12 hours to run out

### Steps to reproduce

N/A

### Minimal reproduction project (MRP)

N/A

**Comment by clayjohn** [comment]
I was reminded of this while working on https://github.com/godotengine/godot/pull/105470 due to some comments on https://www.reddit.com/r/godot/comments/1k0sz07/godot_crashes_after_262k_objects/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button about "RIDs not being freed"

This led me to digging in to how the allocation works in a bit more detail. The following are rough notes about things I learned, I am writing them here to help future-me, or anyone else who reads this:
1. RIDs are allocated on demand and can be freed. 
2. RIDs are basically uint64_ts with two components a 32-bit index into the element pool and a 32 bit "validator"
3. When RIDs are freed, their location in the element pool is available to be re-used and the index is stored in a free list
4. The 32 bit "validator" is a unique number and is masked by `0x7FFFFFFF` so its range is 0-2,147,483,647
5. When doing operations on the RID, we check that the validator contained in the RID matches the one in the element pool
6. The validator is created by incrementing a 64 bit atomic uint (shared by all allocators) and then masking the value to `0x7FFFFFFF`
7. There is a practical limit of 2,147,483,647 unique RIDs because the allocator will crash (intentionally, see https://github.com/godotengine/godot/pull/74759) once that limit is reached

There are a couple of problems with the current approach:
1. The validators will skip values since the atomic uint `base_id` is shared by all allocators. 
2. Similarly, once the sum of all allocation _for all_ allocators combined is `0x7FFFFFFF`, the program will crash. This is irrespective of whether any one allocator has come close to that limit. 

Overall, I am starting to suspect that the validator was originally meant to overflow and wrap around back to 0. 

From the original PR it looks like the validator used a modulo instead of a mask https://github.com/godotengine/godot/commit/4f163972bbd9c7379b01a1f9aa5310646ca7865e#diff-daee6e886b52f0519d3c55ae763f33e09b4176aab3449575c0df05e1ff102e15R79

The modulo was replaced with a mask in https://github.com/godotengine/godot/commit/e38ae8d7e4407722e1b8d2b7caf49052f2f9174b

Looking at that commit, I suspect that randomshaper was mistaken. The rest of the commit fixes the bug, while the change to use a mask instead of modulo was made since it appeared to be a typo (mask is used everywhere else in the code)
